app/models/annotation_merger.rb

```diff
@@ -0,0 +1,117 @@
+class AnnotationMerger
+  def initialize(annotations)
+    # If annotation["text"] is nil, set it to an empty string.
+    # If the text ends with a period, add a space to clarify sentence boundaries when merging annotations.
+    # This prevents loss of readability or meaning.
+    # Setting nil to an empty string also avoids errors in subsequent processing.
+    # Ensure consistent denotations and relations format for reliable merging.
+    @annotations = annotations.map do |annotation|
+      text = annotation["text"] || ""
+      text += " " if text.end_with?(".")
+      denotations = annotation["denotations"] || []
+      relations = annotation["relations"] || []
+      annotation.merge("text" => text, "denotations" => denotations, "relations" => relations)
+    end
+
+    # Pre-build Denotation ID sets
+    denotation_id_sets = @annotations.map { |a| (a["denotations"] || []).map { |d| d["id"] }.to_set }
+
+    # Check referential integrity of relations
+    @annotations.each_with_index do |annotation, idx|
+      relations = annotation["relations"] || []
+      denotation_ids = denotation_id_sets[idx]
+      relations.each do |relation|
+        unless denotation_ids.include?(relation["subj"]) && denotation_ids.include?(relation["obj"])
+          raise ArgumentError, "Relation #{relation.inspect} in chunk #{idx + 1} refers to missing denotation."
+        end
```

Comment:
現在の条件式 `unless denotation_ids.include?(relation["subj"]) && denotation_ids.include?(relation["obj"])`からは、何をチェックしているのかを正確に読み解くのが難しいです。特に、

- 変数名だけでは subj / obj が何を意味するのか分かりづらい
- 条件の意味を理解するためにコードを文脈から逆算する必要がある

という点で、読み手にとって負担の大きいコードになります。

こうした場合、ロジックを意味のあるメソッド名に抽出することで、コードの可読性と意図の明確さが大きく向上します。

たとえば、次のように書くと意図が一目で伝わります：

```suggestion
        unless referable_to?(relation, denotations)
          raise ArgumentError, "Relation #{relation.inspect} in chunk #{idx + 1} refers to missing denotation."
        end
```

まとめると

- 条件式に意味があるなら、名前で表現して抽出
- 特に否定形の unless に複雑な論理を埋め込むと、誤読や見落としの原因になりやすい
- 「意味を持った名前」によって意図が明文化され、保守性も高まる
 
です。

Reply:
https://github.com/jdkim/textae_campus/pull/19/commits/534ea1504b266ce0f1d0ba27e5247af5de10faa3 で修正しました。